### DISCUSSION TITLE
Inactive edge sometimes produce an "active normal"

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#679] [Q&A] [answered]
I wasn't sure whether this could be considered an issue or not, or if I should have just mentioned it in #630 instead, but while looking into another triangle edge problem that I ran into a while back I've now found it to be the result of this early-out in `ActiveEdges::FixNormal`:

https://github.com/jrouwe/JoltPhysics/blob/7a4e43ee91337ba178c2f705c1c5115ea4f28ad9/Jolt/Physics/Collision/ActiveEdges.h#L60-L63

This seems to create some problems with Godot's character controller (in the part that's not implemented on my side) as it uses the contact normals (that I provide it) to figure out what the angle of the floor is and treats the surface as a wall if it exceeds a certain threshold, which defaults to 45 degrees. This means that when you have a character (capsule in this case) walk up a perfectly flat mesh slope that's at an angle of 40-44 (-ish) degrees, and you walk across a triangle edge, you can sometimes end up getting an `inNormal` (rather than `inTriangleNormal`) that exceeds this max floor angle and the character comes to a dead stop.

Removing this early-out resolves this.

I can't help but wonder if Godot's character controller should be using the surface normal rather than the contact normal when comparing this floor angle, like you do with `CharacterBase::IsSlopeTooSteep`, but those are often treated as one and the same in Godot for some reason. This particular problem is also reproducible when using Godot Physics, albeit much less frequently, so there's certainly an argument that this should be fixed from the Godot side of things instead, although it's unclear how readily available the surface normal is in Godot Physics.

Anyway, I was curious to hear what your thoughts were before I create a repro for it, if you feel that one is needed that is.

**Comment by jrouwe** [comment]
Perhaps the 5 degrees is a bit much. I didn't really experiment with this number, we could do some analysis to see what the performance cost of lowering that number to e.g. 1 degree. What I would expect with a perfectly flat mesh slope though is that none of the edges would be active and that it would take the if above and always return the triangle normal:

https://github.com/jrouwe/JoltPhysics/blob/7a4e43ee91337ba178c2f705c1c5115ea4f28ad9/Jolt/Physics/Collision/ActiveEdges.h#L56-L58

W.r.t. using the surface normal rather than the contact normal: That normal is certainly a lot more consistent. You can easily obtain the surface normal just like the CharacterVirtual in your contact collector:

https://github.com/jrouwe/JoltPhysics/blob/7a4e43ee91337ba178c2f705c1c5115ea4f28ad9/Jolt/Physics/Character/CharacterVirtual.cpp#L90-L94

The risk of using the surface normal is that if you have a step and you hit the edge, you might get the vertical plane first and get the red normal and that's why the CharacterVirtual above chooses between surface and contact normal (blue):

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/7126c531-bb92-4515-a491-04e46ed6e177)

After being blocked, CharacterVirtual will enter the stair walk code to avoid suddenly stopping.

If you have knowledge of the higher level system requirements, you can choose between the surface and contact normal based on e.g. movement direction. Or you could even continue collecting contacts and realize that there is another contact in nearly the same spot with a vertical normal and pick that one instead (the trick is in defining 'same spot' as you have no neighboring triangle information).

**Reply by mihe** [reply]
> Perhaps the 5 degrees is a bit much. I didn't really experiment with this number, we could do some analysis to see what the performance cost of lowering that number to e.g. 1 degree.

Would it make sense to make this configurable, like you did with the active/inactive edge threshold, or does it become too much of a footgun perhaps?

> What I would expect with a perfectly flat mesh slope though is that none of the edges would be active and that it would take the if above and always return the triangle normal

Right, yeah, I should have caught that, or just provided a repro in the first place. I didn't realize it mattered until now, but the "slope" in this case is effectively just a triangle strip, so there's almost always at least one active edge.

(The upper right strip in the image is a wall.)

![Slope](https://github.com/jrouwe/JoltPhysics/assets/4884246/287be510-a049-43ea-a049-6d463b5b810e)

> If you have knowledge of the higher level system requirements, you can choose between the surface and contact normal based on e.g. movement direction. Or you could even continue collecting contacts and realize that there is another contact in nearly the same spot with a vertical normal and pick that one instead (the trick is in defining 'same spot' as you have no neighboring triangle information).

While I do have the movement vector, that's pretty much all I have. I don't have the up-vector for the character, which is configurable, so making a choice based on any sort of verticality is going to be tricky. I suppose I could experiment with the same approach you take with `mActiveEdgeMovementDirection` and pick the one that's least obstructing though. That might be worth a shot.

**Reply by jrouwe** [reply]
> Would it make sense to make this configurable, like you did with the active/inactive edge threshold, or does it become too much of a footgun perhaps?

I think that's too much. I'd rather first see if we can just find a better constant.

**Reply by jrouwe** [reply, marked as answer]
I tested it with 1 degree and this has no impact on performance, see #687. Does this fix the issue too? (I guess it can still happen if the floor is at an angle of 44 degrees now, but as a level designer you would really be pushing things)

**Reply by mihe** [reply]
That does seem to fix the issue, or at least alleviate it to such a degree that I struggle to repro it.

While you probably want to merge this anyway, I'm assuming this would be made somewhat redundant by the algorithm proposed in #630, since it would presumably prevent any of these slightly off normals from showing up in the context of the character controller in the first place?

**Reply by jrouwe** [reply]
I think it is indeed a good fix regardless of #630, but I need to test it a bit more first.

**Reply by jrouwe** [reply]
I merged the change.

**Reply by mihe** [reply]
Wonderful, thank you.
